base_delivery_carrier_files/carrier_file.py

```diff
@@ -0,0 +1,153 @@
+        Has to be inherited to add carriers
+        """
+        return [('generic', 'Generic')]
+
+    @api.model
+    def get_write_mode_selection(self):
+        """
+        Selection can be inherited to add more write modes
+        """
+        return [('disk', 'Disk')]
+
+    @api.multi
+    def _write_file(self, filename, file_content):
+        """
+        Method responsible of writing the file, on the filesystem or
+        by inheriting the module, in the document module as instance
+
+        :param browse_record carrier_file: browsable carrier.file
+                                           (configuration)
+        :param tuple filename: name of the file to write
+        :param tuple file_content: content of the file to write
+        :return: True if write is successful
+        """
+        for carrier_file in self:
+            if not carrier_file.export_path:
+                raise exceptions.except_orm(_('Error'),
+                                            _('Export path is not defined '
+                                            'for carrier file %s') %
+                                            (carrier_file.name,))
+            full_path = os.path.join(carrier_file.export_path, filename)
+            with open(full_path, 'w') as file_handle:
+                file_handle.write(file_content)
+        return True
+
+    @api.one
+    def _generate_files(self, picking_ids):
+        """
+        Generate one or more files according to carrier_file configuration
+        for all picking_ids
+
+        :param browse_record carrier_file: browsable carrier file
+                                           configuration
+        :param list picking_ids: list of ids of pickings for which
+                                 we have to generate a file
+        :return: True if successful
+        """
+        # Convert frozendict context to editable context
+        context = dict(self.env.context or {})
+        log = logging.getLogger('delivery.carrier.file')
+        file_generator = new_file_generator(self.type)
+
+        picking_obj = self.env["stock.picking"]
+        pickings = [picking for picking in picking_obj.browse(picking_ids)]
+
+        # must return a list of generated pickings ids to update
+        files = file_generator.generate_files(pickings, self)
+        if self.auto_export:
+            context['picking_id'] = pickings and pickings[0].id
+        for f in files:
+            filename, file_content, picking_ids = f
+            # we pass the errors because the files can still be
+            # generated manually
+            # at first I would like to open a new cursor and
+            # commit the write after each file created
+            # but I encountered lock because the picking
+            # was already modified in the current transaction
+            try:
+                # Instead of replacing everything, append to existing context
+                if self.with_context(**context)._write_file(
```

Comment:
Why copying the whole context to merge it to itself afterwards?
You just want to alter `picking_id` if necessary.

You could do that by writing something like

    self = self.with_context(picking_id=pickings and pickings[0].id)